In [20]:
from sklearn.datasets import fetch_openml
from sklearn.ensemble import RandomForestClassifier
from sklearn.pipeline import make_pipeline
from sklearn.decomposition import PCA
from sklearn.linear_model import SGDClassifier


In [2]:
# Get mnist and split
mnist = fetch_openml('mnist_784', as_frame=True)
X_train, y_train = mnist.data[:60_000], mnist.target[:60_000]
X_test, y_test = mnist.data[60_000:], mnist.target[60_000:]

/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/datasets/_openml.py:1002: FutureWarning: The default value of `parser` will change from `'liac-arff'` to `'auto'` in 1.4. You can set `parser='auto'` to silence this warning. Therefore, an `ImportError` will be raised from 1.4 if the dataset is dense and pandas is not installed. Note that the pandas parser may return different data types. See the Notes Section in fetch_openml's API doc for details.
  warn(


Random forest on MNIST without PCA

In [5]:
rnd_clf = RandomForestClassifier(random_state=42)
rnd_clf.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

In [6]:
rnd_clf.score(X_test, y_test)

0.9705

Random forest on MNIST without PCA:
* Time taken: 24 seconds
* Test accuracy: 97%

Random forest on MNIST after PCA


In [15]:
pca = PCA(n_components=0.95, random_state=42)
X_train_reduced = pca.fit_transform(X_train)

In [16]:
X_train_reduced.shape

(60000, 154)

In [17]:
rnd_clf_reduced = RandomForestClassifier(random_state=42)
rnd_clf_reduced.fit(X_train_reduced, y_train)

RandomForestClassifier(random_state=42)

In [18]:
X_test_reduced = pca.transform(X_test)
rnd_clf_reduced.score(X_test_reduced, y_test)

0.9481

Random forest on MNIST with PCA:
* Time taken: 1.5 minutes
* Test accuracy: 94.8%

Conclusion: PCA not always speeds up the training, it depends on the model and the data...

SGDClassifier on MNIST without PCA

In [21]:
sgd_clf = SGDClassifier(random_state=42)
sgd_clf.fit(X_train, y_train)

SGDClassifier(random_state=42)

In [22]:
sgd_clf.score(X_test, y_test)

0.874

SGDClassifier on MNIST without PCA:
* Time taken: 1.41 minutes
* Test accuracy: 87.4%

SGDClassifier on MNIST with PCA:

In [24]:
sgd_clf = SGDClassifier(random_state=42)
sgd_clf.fit(X_train_reduced, y_train)

SGDClassifier(random_state=42)

In [26]:
sgd_clf.score(X_test_reduced, y_test)

0.8959

SGDClassifier on MNIST with PCA:
* Time taken: 19.9 seconds
* Test accuracy: 89.6%

With SGDClassifier, the training time decreased and the test accuracy increased.